#!/bin/bash

#PBS -l nodes=1:ppn=8,walltime=40:00:00,mem=96gb
#PBS -N gatk_comb
#PBS -m abe
#PBS -M cxb585@psu.edu
#PBS -e localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.e${PBS_JOBID}.${PBS_ARRAYID}
#PBS -o localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.o${PBS_JOBID}.${PBS_ARRAYID}

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------

# Call with, e.g.:
# qsub -A open -t 1-21 pbs/call_gatk_combinedGVCFs.pbs

working_dir=$PBS_O_WORKDIR

module load gatk

GENOME_FA=genomes/rheMac2/rheMac2.fa
GENOME_NAME=rhesus

# ------------------------------------------------------------------------------
# Run program
# ------------------------------------------------------------------------------

cd $working_dir

if [ "$PBS_ARRAYID" -eq 21 ]; then
    CHROM=chrX
else
    CHROM=chr$PBS_ARRAYID
fi

echo "Calling SNPs for chromosome $CHROM"

GVCFS=(`ls results/*.g.vcf`)
count=0

for g in ${GVCFS[*]}; do
    GVCFS[$count]="-V "$g" "
    count=`expr $count + 1`
done

java -Xmx96g -jar ${GATK}/gatk-package-4.1.4.0-local.jar \
    GenomicsDBImport \
    --genomicsdb-workspace-path ${GENOME_NAME}_snps_$CHROM \
    ${GVCFS[*]} \
    -R $GENOME_FA \
    -L $CHROM

exit;
