#!/bin/bash

#PBS -l nodes=1:ppn=8,walltime=2:00:00
#PBS -N fltr_gatk_snps
#PBS -m abe
#PBS -e localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.e${PBS_JOBID}.${PBS_ARRAYID}
#PBS -o localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.o${PBS_JOBID}.${PBS_ARRAYID}

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------

working_dir=$PBS_O_WORKDIR

module load gatk

GENOME_FA=genomes/rheMac2/rheMac2.fa
GENOME_NAME=rhesus

# ------------------------------------------------------------------------------
# Run Program
# ------------------------------------------------------------------------------

cd $working_dir

if [ "$PBS_ARRAYID" -eq 21 ]; then
	CHROM=chrX
else
	CHROM=chr$PBS_ARRAYID
fi

java -Xmx2g -jar ${GATK}/gatk-package-4.1.4.0-local.jar \
    VariantFiltration \
    -R ${GENOME_FA} \
    -O ${GENOME_NAME}_snps/${CHROM}.flt.vcf \
    --variant ${GENOME_NAME}_snps/${CHROM}.raw.snps.indels.vcf \
    --filter-expression "QD < 2.0" \
    --filter-name "QDfilter" \
    --filter-expression "MQ < 40.0" \
    --filter-name "MQfilter" \
    --filter-expression "FS > 60.0" \
    --filter-name "FSfilter"

# Select variants with "FILTER=PASS" and are SNPs:
java -jar ${GATK}/gatk-package-4.1.4.0-local.jar \
	SelectVariants \
	-R ${GENOME_FA} \
	--variant ${GENOME_NAME}_snps/${CHROM}.flt.vcf \
	-select "vc.isNotFiltered() && vc.isSNP()" \
	-O ${GENOME_NAME}_snps/${CHROM}.pass.snp.vcf

exit;
